File: 有效的括号.py


def isValid(s: str):
  # 定义一个栈
  symbol = []
  for i in s:
    if i == "(":
      symbol.append(")")
    elif i == "[":
      symbol.append("]")
    elif i == "{":
      symbol.append("}")
    elif len(symbol) == 0 or i != symbol.pop(): return False
  return len(symbol) == 0


if __name__ == "__main__":
  test = "[]{}()"
  print(isValid(test))

File: 有效的括号.cpp
#include <iostream>
#include <stack>
using namespace  std;


bool isValid(const string& s){
  stack<char> symbol;
  for (int i = 0; i < s.size(); i++){
    if (s[i] == '('){
      symbol.push(')'); continue;
    } 
    if (s[i] == '['){
       symbol.push(']'); continue;
    }
    if (s[i] == '{') {
      symbol.push('}'); continue;
    }
    else if (symbol.empty() || s[i] != symbol.top()) return 0; 
    symbol.pop();
  }
  return symbol.empty();
}

int main(){
  string test = "{[()]}";
  cout << isValid(test) << endl;
  return 0;
}